### ISSUE TITLE
triple click does not select full line

### ISSUE THREAD

**Issue by marc-guenther** [subethaedit/SubEthaEdit#181]
Triple click with the mouse should select the whole line, like it does in every other application.

In SubEtha, it selects some weird part of the line, somehow based on syntax coloring, which is completely confusing.

**Comment by monkeydom** [comment]
This behaves as intended. The additional step is added to accomodate full string selections including the quotation marks. Line just moves on to 4 clicks and whole document is 5 clicks.

**Comment by marc-guenther** [comment]
So confusing the user is intended behaviour?

Sorry, but I am using SubEtha now for over a decade, and still this issue trips me up all the time. I copy something out of a document, and then (sometime even much later) notice, that it copied only parts of what I wanted. This is hugely annoying.

There is something called *consistency* which is a very very important factor in UI behaviour design. Do not surprise the user with unexpected behaviour for commonly known actions. There is a reason why we don't swap the accelerator and brake pedal every Thursday, because it would utterly confuse the user.

Likewise the behaviour of single, double and triple click is hardwired into my brain, it is consistent across applications, and there is absolutely NO reason for a single application to break that consistency and do something unexpected. If you like to select parts of a line, choose some action that isn't already in common use.

When I need to select a line, then I do NOT think, Ah, I have to move my mouse here, and then click it three times. I do not think like this at all. When I need to select a line, then what I think is, ok, I want to select this line, and the unconscious part of my brain moves the mouse and triple clicks. And there is absolutely no way, that this unconscious part of my brain can be trained to think, Ah, this is SubEtha, here I have to click four times. That would make this a conscious action, and would seriously interfere with the thought process of what I really want to be doing at the moment.

Sorry, I don't really know why I must explain this here, as I consider this common sense.

And no thanks for closing this issue without even considering this issue and the reasoning behind it. This is not the first time I reported this. Oh well...

**Comment by sly74fr** [comment]
No offense intended, but I fully agree with Marc on this one. It bothered me since I started using SEE so long ago...

**Comment by monkeydom** [comment]
Well, it might even be us having had the same discussion then yet as there haven't been many other developers on this one, and at this juncture not something I would consider changing lightly, as you have noted, it has been there forever. And it has real world usage and great benefits in other cases. Selecting URLs completely, or MarkdownLinks etc. where there is just nor real alternative. Where selecting the line/paragraph still can be done using either a four click sequence or simply by clicking the line number.

You don't have to explain this, but feel free to make your case. And as you have proven: even a closed issue can still be discussed. At this point the closure is just honest imho as this is not something that we would change lightly, as triple click to select semantic syntax styled entity is a behavior deeply in the DNA of see.

**Comment by sly74fr** [comment]
Could this *technically* become a (hidden) preference ?

**Comment by sly74fr** [comment]
I found some evidences in [SEETextView.m](https://github.com/subethaedit/SubEthaEdit/blob/develop/SubEthaEdit-Mac/Source/SEETextView.m)::selectionRangeForProposedRange() at line 421 but don’t know if it’s circumscribed or more widespread...

**Comment by monkeydom** [comment]
This is the place where it is implemented, yes. I don't like the idea of a preference at all. But I might be swayed to just move it to 4th click and remove the whole document variant, as this is not a system thing (anymore). @marc-guenther would that fix it for you? or are you in the habit to just hammer the mouse 3 or more times to get the line selection?

**Comment by monkeydom** [comment]
If there are strong opinions from others, lets try to collect them here. As we don't have a good feedback loop with the user base currently (maybe also because it is small?), this is a little bit of a tricky one. I know I'm a big proponent of that feature, but it is well hidden, and it is far less essential ever since SubEthaEdit is not a Objective-C external main editor for project builder anymore (which is from a long time ago).

**Comment by monkeydom** [comment]
[SubEthaEdit-5.1.6-9719.zip](https://github.com/subethaedit/SubEthaEdit/files/5559601/SubEthaEdit-5.1.6-9719.zip)
I build a version behaving that way to live on for those interested. It is notarized and fully on par with the web version otherwise.

**Comment by marc-guenther** [comment]
Oh, wow, lot's going on here. :)

Yes, yes, I just downloaded this, and that is exactly what I always wanted all the years. Works perfectly fine. Very happy!

Some comments:
> And it has real world usage and great benefits in other cases.

Yes, of course it has benefits, but that is not the point. I have nothing against selections based on the structure of the document, for example I miss double-click on braces in VSCode to select entire content of the braces.

The point is, that this has been assigned to a user action which already has a different well established meaning, and this is what confuses me every time.

(Whereas implementing double-click on braces is ok, because there is no existing default meaning for that, only for double-click on words, and that didn't change)

> Selecting URLs completely, or MarkdownLinks etc. where there is just nor real alternative.

But there is an alternative, as you have just shown. Quadruple-click :) (and now that I know this feature is there, I might even use it). Adding new features is nice, redefining existing features is bad. But I keep repeating myself :)

(Recently there was a Mac App, which redefined CMD-a (Select-All) to "Add Account". Seriously! So whenever you spent some time in a text field, there was a good chance for this setup assistant to pop up inadvertently. Luckily they very quickly fixed that.)

> ...  as triple click to select semantic syntax styled entity is a behavior deeply in the DNA of see.

That might be true for you as the developer of the software, but definitely not for me as a normal user (I can only speak for myself of course). I MIGHT imagine that now quadruple-click will be deeply imprinted into my DNA after I get used to it, but definitely not triple-click. That was already too deep in my DNA for line selecting.

> would that fix it for you? or are you in the habit to just hammer the mouse 3 or more times to get the line selection?

Yea, of course, that is great. To select a line, I click exactly 3 times, not more. If I would click more often, I wouldn't be selecting a line. ;)

About user feedback. I think most people simply don't care, because:
- they don't use the feature: You would be surprised how many people do not know that something as triple-click even exists. Friend of mine is a professional software developer, he never uses double or single click. Stuff like _double-click-n-drag_ to select multiple words he had never even heard of. There are even people who don't use CMD-c, but use the mouse to select "Copy" from the menu every single time.
- they have no muscle-memory whatsoever: I am completely dependent on that. When I take my keys out of my pocket, I have them n my hand so that from the two identical looking keys, with the left I can open the main door, and the right one opens the apartment. This is not something I know, but rather something that I have done so often, that it's automatic. My friend just tries each key every time.
- they never experienced a consistent OS, and are used to the chaos everywhere: I come from NextStep, where consistency was a key design element. Everything behaved the same everywhere. Find panel, copy pasting formatting styles, main menus, editing text, everything. I once gave a demo of some Nextstep application I wrote at uni, and the first thing they asked, ok, show us the main menu. Ah, why, I wondered, do they want to see that, it's just the main menu, it looks like every other main menu on the system. Then I realized, they come from X11, where the idea of a main menu or any idea of consistency was completely unknown. In X11 you couldn't even copy/paste in a reliable/consistent way. And that was what they were using on a daily basis.

There is a similar issue in Confluence, where double- and triple-clicking selects an entire paragraph instead of words or lines. So people frequently copy entire section of shell commands into their Terminal, when they just wanted to execute one command. This is open for years now, and Atlassian and apparently most users simply don't care.

So, TLDR, thanks for that version of SEE above! Exactly what I needed!

**Comment by sly74fr** [comment]
Yes it fells so much smoother for me too...

I easily understanding your point @monkeydom, and would again advise on the "preference" proposition, in order to keep this feature, but also support standard macOS behaviour if preferred.

**Comment by monkeydom** [comment]
Thanks for your time reillustrating the important points for you on this one. I think it is likely this will make it into the next release for all.

**Comment by sly74fr** [comment]
Thank you so much for your time and all the great work on SEE, my editor of choice since its initial release 👍
